src/mlpack/methods/hdbscan/hdbscan.hpp

```diff
@@ -0,0 +1,218 @@
+/**
+ * @file hdbscan.hpp
+ * @author Sudhanshu Ranjan
+ *
+ * An implementation of the HDBSCAN clustering method.
+ * HDBSCAN works with variable density.
+ *  
+ *
+ * @code
+ * @inproceedings{
+ *   author = {R. J. G. B. Campello, D. Moulavi & Sander, J.},
+ *   title = {{Density-Based Clustering Based
+ *             on Hierarchical Density Estimates.}},
+ * }
+ * @endcode
+ * Reference : "https://github.com/EdwardRaff/JSAT/blob/master/JSAT/
+ *              src/jsat/clustering/HDBSCAN.java"
+ * Reference : "http://hdbscan.readthedocs.io/en/latest/how_hdbscan_works.html"
+ */
+#ifndef __MLPACK_METHODS_HDBSCAN_HDBSCAN_HPP
+#define __MLPACK_METHODS_HDBSCAN_HDBSCAN_HPP
+
+#include <mlpack/core.hpp>
+#include <boost/dynamic_bitset.hpp>
+#include <mlpack/methods/neighbor_search/neighbor_search.hpp>
+#include <mlpack/methods/emst/dtb.hpp>
+#include <mlpack/core/metrics/hdbscan_metric.hpp>
+
+namespace mlpack {
+namespace hdbscan {
+
+template<typename NeighborSearch = neighbor::NeighborSearch
+                                             <neighbor::NearestNeighborSort,
+                                             metric::EuclideanDistance>,
+         typename MetricType = metric::HdbscanMetric,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType = tree::BallTree>
+class HDBSCAN
+{
+ public:
+  /**
+   * Construct the HDBSCAN object with the given parameters.
+   *
+   * @param minPoints The number of points present in a cluster.
+   */
+  HDBSCAN(const size_t minPoints = 10,
+          bool allowSingleCluster = false);
+
+  /**
+   * Performs HDBSCAN clustering on the data,
+   * returning the list of cluster assignments.
+   * The assigned clusters have values from 0 onwards. The noise points  
+   * are labelled as (total nuber of points + 1)
+   *
+   * @param MatType Type of matrix (arma::mat or arma::sp_mat).
+   * @param data Dataset to cluster.
+   * @param assignments Vector to store cluster assignments.
+   */
+  template<typename MatType>
+  void Cluster(const MatType& data,
+               arma::Row<size_t>& assignments);
```

Comment:
I am not sure, but I wanted to see what you thought: do you think that a user might benefit from being able to return a full hierarchy of the clusters?  Let me know what you think.  Maybe it would be some nice functionality.

Reply:
Yes, this is the idea behind Single Linkage Tree, merge the closest clusters to obtain the next cluster and in this way it goes on forming the hierarchy. Should I create a separate method
```
getSLT() 
```

Reply:
I think it would be a good idea, but how about we come back to this point when the rest of the code is ready?  Then the design we should use should be a little clearer.